### DISCUSSION TITLE
Getting the type of variables and their assignment expression details

### DISCUSSION THREAD

**Discussion by yadu-kr** [MikePopoloski/slang#719] [Q&A]
Hi Mike,

I have a simple SV model as shown below:
````
module top
  (
   input wire clk, 
   input wire rstn,
   input wire [31:0] in,
   input wire [31:0] key_in,
   output reg [31:0] out
  );

  reg [31:0] key;
  reg [31:0] sum = 0;
  
  assign key = key_in;

  always
  begin
      sum = key + {in[31], 31'b0};
      out[31:8] = sum[31:8];
	  out[7:0] = in[7:0];
  end
endmodule
````

After adding the syntax tree to a Compilation object and calling the getRoot() function, I'm using the ASTVisitor with the handle() function to traverse the AST and process the nodes. I want to know how I can get the initialization direction of each VariableSymbol (whether input or output). If they are assigned to a value, I also want to get the symbol it is assigned with. 

In the case of the code above, I need to get the assigned symbols `key_in` (for `key`),  `key` and `in` (for `sum`), and `sum` and `in` (for `out`).

My current Visitor class looks like this:

```
struct Visitor : public ASTVisitor<Visitor, true, true> {
    
    // Visitor class 

    template<typename T>
    void handle(const T& t) {
        s_term term;
        std::string p;
        if constexpr (std::is_base_of_v<ValueSymbol, T>) {
            t.getHierarchicalPath(p);               // Get the hierarchical path from current symbol to the root node 
            term.name = p;
            auto width = t.getType().bitstreamWidth();
            if(width)
                term.msb = width-1;
            if constexpr (std::is_base_of_v<NetSymbol, T>) {        // If the symbol is a Net, insert Wire as termtype
                term.termtype.insert("Wire");
            }
            else if constexpr (std::is_base_of_v<VariableSymbol, T>) {          // If the symbol is a Variable, insert Reg as termtype
                term.termtype.insert("Reg");
                
            }
                   

              //  PROCESS ASSIGNMENT EXPRESSIONS 

             

            s_terms[p] = term;                  // Add term to map of terms 

        }
        visitDefault(t);
    }
};
```

It would be helpful if you could give some input. Thanks!

**Comment by MikePopoloski** [comment]
Given a ValueSymbol, you can loop over the ValueDrivers via the `drivers()` method to find all of the expressions that drive a value for that symbol. You should be able to visit those expressions and collect all of the symbols that are used within it. You can also use `getFirstPortBackref()` and walk a linked list of ports that refer to those value symbols to find their directions.

**Reply by yadu-kr** [reply]
Thanks for the reply Mike, I managed to get the port direction using `getFirstPortBackref()`, but I'm not able to iterate over the drivers of the VariableSymbols. Here is the relevant code from my current `handle()` function:

```
    void handle(const T& t) {

    ...

        if constexpr (std::is_base_of_v<ValueSymbol, T>) {
  
            const ValueSymbol::PortBackref* portback = t.getFirstPortBackref();  // Get reference to the attached port of the symbol
            if(portback != nullptr) {
                ArgumentDirection dir = (*portback).port->direction;        // Returns the direction of the corresponding symbol
                if (toString(dir)== "In") {
                    term.termtype.insert("Input");
                }
                else if (toString(dir)== "Out") {
                    term.termtype.insert("Output");
                }   
            }

            .......

            else if constexpr (std::is_base_of_v<VariableSymbol, T>) {     // If the symbol is a Variable, insert Reg as termtype, and add to binds
                term.termtype.insert("Reg");
                bind.dest = pathname;
                for ( auto& driverPtr : t.drivers()) {

                     // GET DRIVERS OF SYMBOLS

                     const Expression* exp = (*driverPtr).prefixExpression;
                     std::cout << t.name << "  " << exp->type << std::endl; 
                }

                s_binds[pathname] = bind; // Add bind to map of binds 
            }

            s_terms[pathname] = term;       // Add term to map of terms 

        }
        visitDefault(t);
    }
```
The statements inside the range-based loop iterating over the drivers are not getting executed at all.

**Reply by MikePopoloski** [reply]
Drivers are only registered with their value symbol when they are first discovered, so you need to visit the whole tree before you can guarantee they've all been found. You can do this prior to running your visitor by doing something like calling getAllDiagnostics() on the compilation, which will force everything to be elaborated and in a finalized state.

**Reply by yadu-kr** [reply]
Got it, that did the trick!

When you said you can collect all the symbols in the driver expressions, how did you mean to do that? I have the driver expressions from the `prefixExpression` pointers inside each `ValueDriver` object. But the `getSymbolReference()` function returns the VariableSymbol that is being driven. I was instead looking for the symbols that are driving the VariableSymbol.

**Reply by MikePopoloski** [reply]
That expression should be an AssignmentExpression (not always I think but most of the time it will be) so you can look at the rhs of that expression. You will need a visitor to visit all of the subexpressions and find ones that are a NamedValueExpression (or HierarchicalValueExpression) since there can be more than one.

**Reply by MikePopoloski** [reply]
Oh sorry, that's wrong. For some reason I thought the driver was tracking the full expression but it appears it's not (because nothing needed to know the rhs as of yet). That's something that could be changed though.

**Reply by yadu-kr** [reply]
Aah okay, so the driver does not include information about the symbols contained in the RHS of the expression?

**Reply by MikePopoloski** [reply]
That's right, currently it does not.

**Reply by yadu-kr** [reply]
Would that be something that could be included in a future release of slang? Maybe I could try and help with the implementation if required.

**Reply by MikePopoloski** [reply]
Yes, I think that's a worthwhile feature to add. I generally want to be careful about how much gets added to ValueDriver since there will be many in a large design and it will contribute heavily to memory usage, but adding one more pointer is probably fine.

**Reply by yadu-kr** [reply]
Understood. Could you point me to where I could start digging around if I wanted to add the feature? 

And while on the topic of tracking the driver expressions, would it be possible to get the operators involved in the driving as well? In the example I gave above, I want to get the `+` operator so that I know `sum` is the addition of `key` and the modified `in`.

**Reply by MikePopoloski** [reply]
You'd need to add it to the ValueDriver class as a member and then populate it everywhere that the drivers are created.

The thing you would track is an Expression object, which is naturally made up of subexpressions and it would be easy to look at them to know which operators are involved.
